File: server/models/report.go
package models

import (
	"fmt"
	"strings"
	"unicode/utf8"
)

type ReportReason string

const (
	ReportReasonSpam          ReportReason = "spam"
	ReportReasonHarassment    ReportReason = "harassment"
	ReportReasonInappropriate ReportReason = "inappropriate_content"
	ReportReasonImpersonation ReportReason = "impersonation"
	ReportReasonOther         ReportReason = "other"
)

var validReportReasons = map[ReportReason]bool{
	ReportReasonSpam:          true,
	ReportReasonHarassment:    true,
	ReportReasonInappropriate: true,
	ReportReasonImpersonation: true,
	ReportReasonOther:         true,
}

type ReportStatus string

const (
	ReportStatusPending   ReportStatus = "pending"
	ReportStatusReviewed  ReportStatus = "reviewed"
	ReportStatusResolved  ReportStatus = "resolved"
	ReportStatusDismissed ReportStatus = "dismissed"
)

// Report — time fields are strings because modernc.org/sqlite doesn't auto-convert to time.Time.
type Report struct {
	ID             string             `json:"id"`
	ReporterID     string             `json:"reporter_id"`
	ReportedUserID string             `json:"reported_user_id"`
	Reason         ReportReason       `json:"reason"`
	Description    string             `json:"description"`
	Status         ReportStatus       `json:"status"`
	ResolvedBy     *string            `json:"resolved_by"`
	ResolvedAt     *string            `json:"resolved_at"`
	CreatedAt      string             `json:"created_at"`
	Attachments    []ReportAttachment `json:"attachments"`
}

// ReportAttachment — evidence file (images only). Parallel structure to Attachment/DMAttachment.
type ReportAttachment struct {
	ID        string  `json:"id"`
	ReportID  string  `json:"report_id"`
	Filename  string  `json:"filename"`
	FileURL   string  `json:"file_url"`
	FileSize  *int64  `json:"file_size"`
	MimeType  *string `json:"mime_type"`
	CreatedAt string  `json:"created_at"`
}

// ReportWithUsers — report with reporter and reported user info for admin panel.
type ReportWithUsers struct {
	Report
	ReporterUsername string  `json:"reporter_username"`
	ReporterDisplay  *string `json:"reporter_display_name"`
	ReportedUsername string  `json:"reported_username"`
	ReportedDisplay  *string `json:"reported_display_name"`
}

var validReportStatuses = map[ReportStatus]bool{
	ReportStatusPending:   true,
	ReportStatusReviewed:  true,
	ReportStatusResolved:  true,
	ReportStatusDismissed: true,
}

type UpdateReportStatusRequest struct {
	Status string `json:"status"`
}

func (r *UpdateReportStatusRequest) Validate() error {
	status := ReportStatus(r.Status)
	if !validReportStatuses[status] {
		return fmt.Errorf("invalid report status: %s", r.Status)
	}
	return nil
}

type CreateReportRequest struct {
	Reason      string `json:"reason"`
	Description string `json:"description"`
}

func (r *CreateReportRequest) Validate() error {
	r.Description = strings.TrimSpace(r.Description)

	reason := ReportReason(r.Reason)
	if !validReportReasons[reason] {
		return fmt.Errorf("invalid report reason: %s", r.Reason)
	}

	descLen := utf8.RuneCountInString(r.Description)
	if descLen < 10 {
		return fmt.Errorf("description must be at least 10 characters")
	}
	if descLen > 1000 {
		return fmt.Errorf("description must be at most 1000 characters")
	}

	return nil
}

File: client/src/api/report.ts
/**
 * Report API — user reporting endpoint.
 *
 * Supports multipart (with evidence files) or JSON body.
 */

import { apiClient } from "./client";

export type ReportReason =
  | "spam"
  | "harassment"
  | "inappropriate_content"
  | "impersonation"
  | "other";

export type CreateReportRequest = {
  reason: ReportReason;
  description: string;
};

export type ReportAttachment = {
  id: string;
  report_id: string;
  filename: string;
  file_url: string;
  file_size: number | null;
  mime_type: string | null;
  created_at: string;
};

export type Report = {
  id: string;
  reporter_id: string;
  reported_user_id: string;
  reason: ReportReason;
  description: string;
  status: string;
  created_at: string;
  attachments: ReportAttachment[];
};

/** Reports a user. Uses multipart/form-data when evidence files are provided. */
export function reportUser(userId: string, req: CreateReportRequest, files?: File[]) {
  if (files && files.length > 0) {
    const formData = new FormData();
    formData.append("reason", req.reason);
    formData.append("description", req.description);
    for (const file of files) {
      formData.append("files", file);
    }
    return apiClient<Report>(`/users/${userId}/report`, {
      method: "POST",
      body: formData,
    });
  }

  return apiClient<Report>(`/users/${userId}/report`, {
    method: "POST",
    body: req,
  });
}
